flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/StronglyConnectedComponentsComputeUtils.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.flink.runtime.executiongraph.failover.flip1;
+
+import org.apache.flink.api.java.tuple.Tuple2;
+
+import java.util.ArrayDeque;
+import java.util.Arrays;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Utility for computing strongly connected components.
+ *
+ * <p>The computation is an implementation of Tarjan's algorithm.
+ *
+ * <p>Ref: https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm.
+ */
+public final class StronglyConnectedComponentsComputeUtils {
+
+	static Set<Set<Integer>> computeStronglyConnectedComponents(final int numVertex, final List<List<Integer>> outEdges) {
+		final Set<Set<Integer>> stronglyConnectedComponents = new HashSet<>();
+
+		final int[] vertexIndices = new int[numVertex];
+		Arrays.fill(vertexIndices, -1);
+
+		final int[] vertexLowLinks = new int[numVertex];
+		final Deque<Integer> stack = new ArrayDeque<>(numVertex);
+		final boolean[] onStack = new boolean[numVertex];
+		final AtomicInteger indexCounter = new AtomicInteger(0);
+
+		for (int vertex = 0; vertex < numVertex; vertex++) {
+			if (vertexIndices[vertex] == -1) {
+				dfs(vertex, outEdges, vertexIndices, vertexLowLinks, stack, onStack, indexCounter, stronglyConnectedComponents);
+			}
+		}
+
+		return stronglyConnectedComponents;
+	}
+
+	private static void dfs(
+			final int rootVertex,
+			final List<List<Integer>> outEdges,
+			final int[] vertexIndices,
+			final int[] vertexLowLinks,
+			final Deque<Integer> stack,
+			final boolean[] onStack,
+			final AtomicInteger indexCounter,
+			final Set<Set<Integer>> stronglyConnectedComponents) {
+
+		final Deque<Tuple2<Integer, Integer>> loopStack = new ArrayDeque<>();
+		loopStack.add(new Tuple2<>(rootVertex, 0));
+
+		while (!loopStack.isEmpty()) {
+			Tuple2<Integer, Integer> tuple = loopStack.pollLast();
+			final int currentVertex = tuple.f0;
+			final int vertexLoopCount = tuple.f1;
+
+			if (vertexLoopCount == 0) {
+				vertexIndices[currentVertex] = indexCounter.get();
```

Comment:
I would add more comments for respective code blocks or maybe even additionally break the loop body into three methods/steps:
```
startTraversingVertex
finishTaversingOutEdge
traverseOutEdges
createConnectedComponent
```